### DISCUSSION TITLE
How can I get consistent results when getting CPU load between ticks for descendant processes ?

### DISCUSSION THREAD

**Discussion by facewindu** [oshi/oshi#2650] [Q&A] [answered]
I'm doing sth like
```
class Foo {

    private final CentralProcessor centralProcessor;
    private final int pid;
    private final OperatingSystem os;
    private final Map<Integer, OSProcess> oldChildProcessInfo = new HashMap<>();

    public Foo(OperatingSystem os, CentralProcessor centralProcessor) {
        this.os = os;
        this.centralProcessor = centralProcessor;
        this.pid = os.getCurrentProcess().getProcessID();
    }

    public void capture() {
        double descendantsLoad = 0;
        List<OSProcess> descendantProcesses = os.getDescendantProcesses(pid, OperatingSystem.ProcessFiltering.VALID_PROCESS, null, 0);
        for (OSProcess descendantProcess : descendantProcesses) {
            double descLoad = descendantProcess.getProcessCpuLoadBetweenTicks(oldChildProcessInfo.get(descendantProcess.getProcessID()));
            System.out.println(descendantProcess.getProcessID() + " descendant load: " + descLoad + ", state: " + descendantProcess.getState() + ", threads: " + descendantProcess.getThreadCount());
            descendantsLoad += descLoad;
        }
        double ratio = descendantsLoad / centralProcessor.getLogicalProcessors().size();
        if (ratio < 0 || ratio > 1) {
            System.out.println("Invalid my process descendants load value: " + descendantsLoad + " over " + centralProcessor.getLogicalProcessors().size() + " processors");
        }
        oldChildProcessInfo.clear();
        for (OSProcess descendantProcess : descendantProcesses) {
            oldChildProcessInfo.put(descendantProcess.getProcessID(), descendantProcess);
        }
    }
}
```

Where I'm calling `Foo#capture` from a `Executors.newScheduledThreadPool`, every 500ms (I tried doing it every 5s but the problem still happens).
I'm on latest MacOs Sonoma, but I also see the problem on Windows, possibly Linux.

My problem is that the output I see is
```
18546 descendant load: 1.9257703081232493, state: WAITING, threads: 33
18545 descendant load: 1.9953407105416423, state: WAITING, threads: 33
18544 descendant load: 2.394920868605079, state: WAITING, threads: 44
18543 descendant load: 2.3896068928379104, state: WAITING, threads: 42
18542 descendant load: 2.852354688162919, state: WAITING, threads: 44
18541 descendant load: 3.4737133370226894, state: WAITING, threads: 38
Invalid my process descendants load value: 15.031706805293489 over 12 processors
```

My descendant processes are all under heavy stress (this is in a test I'm writing).

If I sum all their individual loads, I get `15.031706805293489`, and IIUC the javadoc, I need to divide by `centralProcessor.getLogicalProcessors().size()` to get a proper value between 0 and 1 (that I can convert to a percentage).
As you see, this is not the case here, as I have 12 CPUs, and 15 / 12 > 1

Am I doing sth wrong ?

**Comment by facewindu** [comment]
If I get this right, I can reason about
- `CentralProcessor#getSystemCpuLoadBetweenTicks` being a value between 0 and1
- my process CPU load (`OperatingSystem#getCurrentProcess#getProcessCpuLoadBetweenTicks`)
- and my descendant processes loads (as done above to get a single value in [0,1]

Basically summing the 'load' from my process and all its descendants should never be above the current system CPU load returned by oshi, correct ?

**Reply by dbwiddis** [reply]
I see you instantiating an `oldChildProcessInfo` map, but it's empty (you never save it) so you're passing `null` to the between-ticks calculation.

This means you're getting the cumulative (since start time) load of each process.   Since each process may have started at a different time, the values you're averaging may represent different time intervals.
 - As a simplified example of what you're probably seeing, imagine a 2-core machine with 2 processes each consuming "max CPU".  Process 1 starts at time 0 and uses both processors so has 200% load.   Process 2 starts at time 500ms, so both processes are using 100% of one of the pocessors and both have a 100% load.  At time 1s, the average usage of process 1 over its 1-second lifespan is 150%, while the average usage of process 2 is 100% (for the 500ms it was up).  This would give you 250% if you summed up these long term averages, or 125% if you divided by the two processors.

I think you need to add a line to your code to save each new `OSProcess` in the map so the "between ticks" calculation is across the same polling interval.  You still may end up with some processes that have just started up within that interval, and miss some that have shut down, so the results may still be slightly off, but should at least (I expect) be less than 1.

Another thing to consider is the tick counter resolution.  Shouldn't be an issue on macOS but on Windows, ticks have 1/64-second (15.625 ms) resolution, meaning there's only 32 (or 31 or 33) ticks counted in a 500-ms polling interval while the "up time" denominator may have a different 32 (or 31 or 33) number of ticks.  That's up to a +/- 6.5% inaccuracy per-process, per-processor.  Your method would average over the processes to mitigate this, but adding up the total processors does not necessarily average so you could have up to 77% inaccuracy in the total active ticks (a process using 100% of all CPUs could measure from 23% to 177% in a 500ms interval.)

**Reply by facewindu** [reply]
I'm sorry, I forgot to paste the end of the code, where the `oldChildProcessInfo` map is updated. I edited the snippet at the top.

The 'new' processes might not be in the old map, if they started after the last polling time. Then their cumulative load is used (since their start time) and this might give them a higher load than if I was comparing their kernel+user time over the full stretch of the polling interval. Is that what you mean?

If that's the case, I can craft some code to compute the ratio myself, and not rely on the cumulative load for those newborn processes.

**Reply by dbwiddis** [reply]
If you're properly updating the map, then the very first polling may be wrong (for reasons cited above) but the subsequent checks should be valid.   (You could work around this by ignoring the first "capture" event results.). Is this the case or do you still have the problem after the first initialization?

**Reply by facewindu** [reply]
ok I think I found the solution, ensuring all process loads are computed on the same tick, which I take as the `max(startTime+upTime)` (aka "captureTime" for the rest) for all processes I'm grabbing.
In the "old process map", I store 'kernel+userTime' and "captureTime".

When iterating on the next capturing batch, processes that are present are
- either new from the last capturing time. For them I get the cumulative load, but interpolate it from the old capturing time (so I keep a lower value)
- or they were already in the last capturing batch, and I grab the kernel + user time I stored for them, then also do a `max(startTime+upTime)` for the new batch, and use all that for computing the new load.

**Reply by facewindu** [reply]
so basically the solution is to ensure the processes are always checked at the same time interval, to ensure the load that are summed up are about the exact same interval.

**Reply by dbwiddis** [reply, marked as answer]
> When iterating on the next capturing batch, processes that are present are
> 
> * either new from the last capturing time. For them I get the cumulative load, but interpolate it from the old capturing time (so I keep a lower value)
> * or they were already in the last capturing batch, and I grab the kernel + user time I stored for them, then also do a `max(startTime+upTime)` for the new batch, and use all that for computing the new load.

For processes that were already present, shouldn't need to do any calculations.  The "Between Ticks" calculation should work just fine, using the difference in up time between the two samples.

The only special handling you need are for processes that have just started between the ticks (in which case you'd want to scale the calculated ticks to the entire interval, e.g. if it used 50% for half the interval, you'd get a cumulative 50%, but in reality you're only using 25% of the total CPU for that interval), or ended during that interval (you have no data to measure so you're going to miss data here.)

**Reply by facewindu** [reply]
all good
